Fix Stats information fails to load when date cannot be parsed when returned from API

staskus opened the pull request:
Fixes #22859

- [ ] Make sure WPKit Podspec is correctly updated

We need to pass `yyyy-MM-dd` date to `API` when fetching `stats/visits` and other similar endpoints. However, in some cases, date conversions to date string fails. 

The bug is related to TimeZone and DTS. For example, Paraguay has a DTS change at midnight 2023-10-01 which is ambiguous to `DateFormatter` and `nil` is returned.

Date formatter we use:

```
private var periodDataQueryDateFormatter: DateFormatter {
    let df = DateFormatter()
    df.locale = Locale(identifier: "en_US_POSIX")
    df.dateFormat = "yyyy-MM-dd"
    return df
}
```

Result for "America/Asuncion" time zone:

```
df.timeZone = TimeZone(identifier: "America/Asuncion")
df.date(from: "2023-10-01") // result nil
```

Result for "Europe/Vilnius" time zone:

```
po df.timeZone = TimeZone(identifier: "Europe/Vilnius")
po df.date(from: "2023-10-01") // result <Date> 2023-09-30 21:00:00 +0000
 ```

[In documentation](https://developer.apple.com/documentation/foundation/dateformatter) Apple suggests to set the timezone to UTC:

> When working with fixed format dates, such as RFC 3339, you set the [dateFormat](https://developer.apple.com/documentation/foundation/dateformatter/1413514-dateformat) property to specify a format string. For most fixed formats, you should also set the [locale](https://developer.apple.com/documentation/foundation/dateformatter/1411973-locale) property to a POSIX locale ("en_US_POSIX"), and set the [timeZone](https://developer.apple.com/documentation/foundation/dateformatter/1411406-timezone) property to UTC.

> RFC3339DateFormatter.timeZone = TimeZone(secondsFromGMT: 0)

However, this is not a straightforward solution, since in Stats we're working with dates that are in a different time zone.

### Solution

Created `RFC339NoTimeDateFormatter` that sets:

```
df.locale = Locale(identifier: "en_US_POSIX")
df.dateFormat = "yyyy-MM-dd"
df.timeZone = TimeZone(secondsFromGMT: 0)
```

but also recalculates the data to the current time zone after conversion. Therefore, the dates remain unchanged but conversions also succeed. I added `RFC339NoTimeDateFormatterTests` as an example.

### To test:

The easiest way to reproduce and confirm that the issue is fixed is to follow the steps from https://github.com/wordpress-mobile/WordPress-iOS/issues/22859

1. Set defaultPeriodCount from 14 to 50 in `SiteStatsTableHeaderView` view so we can go back further back in dates
1. Switch the time zone to Paraguay/Asuncion
2. Open Stats Traffic tab
3. Select weeks
4. Come back to Sep 25 - Oct 1, 2023 week
5. Confirm data loads

## Video before the fix

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/8fd8e053-5fad-4525-9726-df274ad06b76

## Video after the fix

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/582f7075-8bee-45d4-b4a3-54e491ad9a67

## Regression Notes
1. Potential unintended areas of impact

There shouldn't be any. WPKit-PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/771 adds unit tests to make sure that a new date formatter behaves in exact same way but additionally doesn't fail during DST changes.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Unit tests, manual tests

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented:
Update after https://github.com/wordpress-mobile/WordPress-iOS/pull/23366 is merged

staskus committed:
Create RFC339NoTimeDateFormatter

RFC339NoTimeDateFormatter creates yyyy-MM-dd date from string while taking into account DST (Daylight Savings Time)

staskus committed:
Use RFC339NoTimeDateFormatter for some of the Stats endpoints

staskus committed:
Update WordPressKit.podspec

staskus commented:
This is a fix for a low-priority corner case. I only found this bug to be reproducible for Paraguay/Asuncion time zones during daylight savings date. 🤔 Therefore, I'm conflicted about whether it merits significant date formatter updates, although I tried to be careful and cover corner cases with tests.

staskus committed:
Update StatsPostingStreakInsight to use RFC339NoTimeDateFormatter

kean commented:
Dropping a note here as well, as it was discussed internally: it might be worth switching to [native date](https://github.com/CreateAPI/NaiveDate/blob/97176a265165680313f4c3a34e98ab1303932912/Sources/NaiveDate.swift#L6) here to eliminate any potential issues with timezones, time, and `Date` conversions.

staskus commented:
I'm closing this PR based [on the previous discussion](https://github.com/wordpress-mobile/WordPress-iOS/pull/22910#issuecomment-2183280692). This is an extremely rare corner case that only happens when parsing a specific date in a timezone on which DST starts at midnight. Other more generic solutions are preferred.
